src/com/goide/inspections/GoPlaceholderChecker.java

```diff
@@ -0,0 +1,497 @@
+
+    GoExpression placeholder = arguments.get(placeholderPosition);
+    if (!isStringPlaceholder(placeholder)) {
+      String message = "Value used for formatting text does not appear to be a string";
+      myHolder.registerProblem(placeholder, message, ProblemHighlightType.GENERIC_ERROR_OR_WARNING);
+      return;
+    }
+
+    String placeholderText = resolve(placeholder);
+    if (placeholderText == null) return;
+
+    // We always receive the text with double quotes at the beginning and at the end so remove them
+    placeholderText = placeholderText.substring(1, placeholderText.length() - 1);
+
+    if (!hasPlaceholder(placeholderText) && arguments.size() > placeholderPosition) {
+      String message = "Value used for formatting text does not appear to contain a placeholder";
+      myHolder.registerProblem(placeholder, message, ProblemHighlightType.GENERIC_ERROR_OR_WARNING);
+      return;
+    }
+
+    int firstArg = placeholderPosition + 1;
+    int argNum = firstArg;
+    boolean indexed = false;
+    int w;
+    for (int i = 0; i < placeholderText.length(); i += w) {
+      w = 1;
+      if (placeholderText.charAt(i) == '%') {
+        FormatState state = parsePrintfVerb(placeholder, placeholderText.substring(i), firstArg, argNum);
+        if (state == null) return;
+        w = state.format.length();
+        if (state.indexed) indexed = true;
+
+        // One error per format is enough
+        if (!okPrintfArg(placeholder, arguments, state)) return;
+
+        if (!state.argNums.isEmpty()) {
+          int maxArgNum = Collections.max(state.argNums);
+          if (!indexed) maxArgNum += 1;
+          if (argNum < maxArgNum) argNum = maxArgNum;
+        }
+      }
+    }
+
+    if (hasVariadic(myCallExpr.getArgumentList()) && argNum >= arguments.size() - 1) {
+      return;
+    }
+
+    if (!indexed && argNum != arguments.size()) {
+      int expect = argNum - firstArg;
+      int numArgs = arguments.size() - firstArg;
+      String message = String.format("Got %d placeholder(s) for %d arguments(s)", expect, numArgs);
+      myHolder.registerProblem(placeholder, message, ProblemHighlightType.GENERIC_ERROR_OR_WARNING);
+    }
+  }
+
+  // The types of expressions a printf verb accepts. It is a bitmask.
+  protected enum PrintfArgType {
+    argBool(1),
+    argInt(2),
+    argRune(3),
+    argString(4),
+    argFloat(5),
+    argComplex(6),
+    argPointer(7),
+    anyType(-1);
+
+    private int myVal;
+
+    PrintfArgType(int val) {
+      myVal = val;
+    }
+
+    public int getValue() {
+      return myVal;
+    }
+  }
+
+
+  private static class PrintVerb {
+    public char verb;
+    public String flags;
+    public int type;
+
+    public PrintVerb(char verb, String flags, int type) {
+      this.verb = verb;
+      this.flags = flags;
+      this.type = type;
+    }
+  }
+
+  @NotNull private static List<PrintVerb> printVerbs = ContainerUtil.newArrayList(
```

Comment:
why not final? are you going to reassign it?

why not enum? are you going to change inner values?